Tryouts/Plugins/ApplicationPlugins/process explorer/backend/ModuleProcessMonitor/Processes/ProcessInfoGeneratorWindows.cs

```diff
@@ -0,0 +1,360 @@
+﻿// Morgan Stanley makes this available to you under the Apache License,
+// Version 2.0 (the "License"). You may obtain a copy of the License at
+// 
+//      http://www.apache.org/licenses/LICENSE-2.0.
+// 
+// See the NOTICE file distributed with this work for additional information
+// regarding copyright ownership. Unless required by applicable law or agreed
+// to in writing, software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions
+// and limitations under the License.
+
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Management;
+using System.Runtime.InteropServices;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using ModuleProcessMonitor.Logging;
+
+namespace ModuleProcessMonitor.Processes;
+
+[System.Runtime.Versioning.SupportedOSPlatform("windows")]
+internal class ProcessInfoGeneratorWindows : ProcessInfoManager
+{
+    private readonly ILogger<ProcessInfoManager> _logger;
+    private readonly ConcurrentDictionary<int, PerformanceCounter> _cpuPerformanceCounters;
+    private readonly ConcurrentDictionary<int, PerformanceCounter> _memoryPerformanceCounters;
+    private ManagementEventWatcher? _watcher;
+    private readonly object _lock = new();
+    public ProcessInfoGeneratorWindows(ILogger<ProcessInfoManager>? logger)
+    {
+        _logger = logger ?? NullLogger<ProcessInfoGeneratorWindows>.Instance;
+        _cpuPerformanceCounters = new();
+        _memoryPerformanceCounters = new();
+    }
+
+    public ProcessInfoGeneratorWindows(EventHandler<int> sendNewProcess,
+        EventHandler<int> sendTerminatedProcess, EventHandler<int> sendModifiedProcess,
+        ILogger<ProcessInfoManager>? logger = null)
+        : this(logger)
+    {
+        base.SetEvents(sendModifiedProcess, sendNewProcess, sendTerminatedProcess);
+    }
+
+    public override int? GetParentId(Process? process)
+    {
+        if (process == null)
+        {
+            return null;
+        }
+
+        int ppid = 0;
+
+        using ManagementObject mo =
+            new ManagementObject(
+                $"win32_process.handle='{process.Id}'");
+        try
+        {
+            mo.Get();
+            ppid = Convert.ToInt32(mo["ParentProcessId"]);
+        }
+        catch (Exception exception)
+        {
+            if (process.Id > 0)
+            {
+                _logger.ManagementObjectPpid(process.Id, exception);
+            }
+        }
+
+        return ppid;
+
+    }
+
+    public override float GetMemoryUsage(Process process)
+    {
+        int memsize;
+
+        var perf = new PerformanceCounter()
```

Comment:
This way you are always creating a new PerformanceCounter (and you never dispose it).
There is an overload of GetOrAdd that uses a Func for creating the value if necessary, that would be a better choice in this case.

Reply:
You are totally right. I will change them to ....GetOrAdd(intValue, _ => {creating performanceCounter})